fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryLockFactory.java

```diff
@@ -222,7 +192,26 @@ protected void fileLockClearIfLocked() {
                 agilityContext.asyncToSync(LuceneEvents.Waits.WAIT_LUCENE_FILE_LOCK_CLEAR,
                         agilityContext.apply(fileLockFunc));
             }
-            flushAndClose();
+            // always flush before declaring this object as closed. Else agilityContext may fail to commit, while this lock may skip retry
+            agilityContext.flush();
+            closed = true;
+        }
+
+        protected void fileLockClearIfLocked() {
```

Comment:
Can you add a test to `FDBDirectoryLockTest` or higher that fails without this change, and succeeds with it.
Based on the description, I think you should be able to get a better reproduction of the problem by adding to `FDBDirectoryTest` and creating an `FDBDirectory` with an agile context, and simulating what you think is happening. Potentially closing the `AgileContext` itself rather than having a timeout occur.

Reply:
✅